### PR TITLE
AB#74319 : Save records as draft

### PR DESCRIPTION
**Pull request by LenaelleL**
# Description

Adds a new feature that allows the user to save draft records and select them to fill a form again.
User story 74319 
Ticket 77590
Not a 100% sure I added JSCode everywhere...Sorry

## Useful links

- Please insert link to ticket : https://dev.azure.com/WHOHQ/EMSSAFE/_workitems/edit/74319
- Please insert link to back-end branch if any: https://github.com/ReliefApplications/ems-backend/pull/802

## Type of change

Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Improvement (refactor or addition to existing functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# How Has This Been Tested?

- [x] Tested in back-office with different forms

## Screenshots

image

image
image
image

# Checklist:

( * == Mandatory ) 

- [x] * I have set myself as assignee of the pull request
- [x] * My code follows the style guidelines of this project
- [x] * Linting does not generate new warnings
- [x] * I have performed a self-review of my own code
- [x] * I have put the ticket for review, adding the oort-frontend team to the list of reviewers
- [x] * I have commented my code, particularly in hard-to-understand areas
- [x] * I have put JSDoc comment in all required places
- [x] * My changes generate no new warnings
- [x] * I have included screenshots describing my changes if relevant
- [x] * I have selected labels in the Pull Request, according to the changes with code brings
- [ ] I have made corresponding changes to the documentation ( if required )
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [ ] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

**Review by estelafs — APPROVED** [review]
Everything working as expected, i just did some small improvements:
Added missing JSDocs and translations, removed console logs and unused imports, added data checks and added the empty template when table don't have drafts to display

Also, @AntoineRelief , the draft feature should be available in the here in the shared-form-modal when adding or updating a record from a grid?

Screenshot from 2023-10-31 17-05-19

File: libs/shared/src/i18n/en.json [hunk 1]
```diff
@@ -723,6 +723,25 @@
 			"display": {
 				"submissionMessage": "The form has successfully been submitted."
 			},
+			"draftRecords": {
+				"buttonLoad": "Load draft record",
+				"confirmModal": {
+					"confirmDelete": "Do you really want to delete this draft record ?",
```
Comment (AntoineRelief) [inline libs/shared/src/i18n/en.json]: @LenaelleRelief @estelafs 
for your own knowledge, no space between the sentence and the "?" in english ( I don't like that, but that's how it is 😅 )

File: libs/shared/src/lib/components/draft-record-modal/draft-record-modal.component.ts [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+  templateUrl: './draft-record-modal.component.html',
+  styleUrls: ['./draft-record-modal.component.scss'],
+})
+export class DraftRecordModalComponent implements OnInit {
+  /** Data */
+  public loading = true;
+  public survey!: SurveyModel;
+
+  /**
+   * Creates an instance of DraftRecordModalComponent.
```
Comment (AntoineRelief) [inline libs/shared/src/lib/components/draft-record-modal/draft-record-modal.component.ts]: @LenaelleRelief cc @estelafs 
the comment there is not good
constructor must have a similar jsdoc description to the component itself

File: libs/shared/src/lib/models/draftRecord.model.ts [hunk 3]
```diff
@@ -0,0 +1,31 @@
+import { Form } from './form.model';
+import { User } from './user.model';
+
+/** Model for version attributes. */
+interface Version {
```
Comment (AntoineRelief) [inline libs/shared/src/lib/models/draftRecord.model.ts]: @LenaelleRelief 
cc @estelafs 
for file names, prefer to use this format:
"draft-record", than this one: "draftRecord"

File: libs/shared/src/lib/components/form/form.component.ts [hunk 4]
```diff
@@ -321,6 +403,22 @@ export class FormComponent
     }
   }
 
+  /**
+   * Opens the modal to select a draft record to fill the form
+   */
+  public openSelectDraftRecord(): void {
+    const dialogRef = this.dialog.open(SelectDraftRecordModalComponent, {
```
Comment (AntoineRelief) [inline libs/shared/src/lib/components/form/form.component.ts]: @LenaelleRelief 
cc @estelafs 
invalid import
now, for modals, we prefer to use lazy loading import

File: libs/shared/src/lib/components/select-draft-record-modal/graphql/queries.ts [hunk 5]
```diff
@@ -0,0 +1,24 @@
...
+// === GET DRAFT RECORDS ===
+
+/** Graphql request for getting draft records */
+export const GET_DRAFT_RECORDS = gql`
+  query GetDraftRecords($formId: ID!) {
+    draftRecords(formId: $formId) {
+      id
+      createdAt
+      data
+      form {
```
Comment (AntoineRelief) [inline libs/shared/src/lib/components/select-draft-record-modal/graphql/queries.ts]: @LenaelleRelief cc @estelafs 
note that when doing that you query the form once per draft record, so it's a loooot of duplications

File: libs/shared/src/lib/components/form/form.component.html [hunk 6]
```diff
@@ -1,15 +1,17 @@
 <!-- Form header -->
 <div class="flex flex-wrap justify-between items-center gap-4 mb-4">
-  <shared-record-summary
```
Comment (AntoineRelief) [inline libs/shared/src/lib/components/form/form.component.html:3]: @LenaelleRelief cc @estelafs 
I didn't ask to remove that
I'll revert it

Comment (LenaelleL) [reply libs/shared/src/lib/components/form/form.component.html:3]: Sorry, I must have misunderstood !

**Review by AntoineRelief — CHANGES_REQUESTED** [review]
a few things that don't work as expected:
- in the draft record modal ( when previewing ) we don't see the pages
- I didn't check all in details, but I think some methods that were used to remove temporary records after saving ( for resource / resources questions ) were disabled -> to double check
- the possibility to see draft should also be visible in the modal when adding new records ( from a grid, for example ). It may need to refactor the code, so the button to open the draft is a component ( instead of being a button in the form.component as it is now ).
- when the draft is saved ( not as a draft, but as a record ), the draft is not removed. It should be

**Comment by estelafs** [comment]
Done:

- Added the page tabs and navigations to the draft record modal (when previewing)
- Created the shared DraftRecordComponent to add the draft button also in the form and record modal (using the form-builder and the form-helper services for the common methods)
- Now when a draft is loaded and saved as a new record, the draft used is deleted
- I started updating the ResourceModalComponent and the survey utils.ts file to always create draft files instead of temporary (not finished)

**Comment by estelafs** [comment]
Done: updated ResourceModalComponent and the survey utils.ts file to always create draft files instead of temporary records with the local storage

**Comment by AntoineRelief** [comment]
@LenaelleRelief  @estelafs 
really good combined work 👏👏👏
that's a great new feature

**Pull request merged**